File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    static class Node {
        int number;
        int weight;
        
        public Node(int number, int weight) {
            this.number = number;
            this.weight = weight;
        }
    }
    
    public int solution(int N, int[][] road, int K) {
        List<Node>[] graph = createGraph(N, road);
        return (int) Arrays.stream(calcualteWeight(graph))
            .filter(value -> value <= K)
            .count();
    }
    
    private List<Node>[] createGraph(int N, int[][] road) {
        List<Node>[] graph = new List[N+1];
        for(int i = 0; i < N + 1; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < road.length; i++) {
            graph[road[i][0]].add(new Node(road[i][1], road[i][2]));
            graph[road[i][1]].add(new Node(road[i][0], road[i][2]));
        }
        
        return graph;
    }
    
    private int[] calcualteWeight(List<Node>[] graph) {
        int[] weights = new int[graph.length];
        Arrays.fill(weights, Integer.MAX_VALUE);
        weights[1] = 0;
        Node startNode = new Node(1,0);
        
        Queue<Node> queue = new PriorityQueue<>(new Comparator<Node>() {
            @Override
            public int compare(Node node1, Node node2) {
                return node1.weight - node2.weight;
            }
        });
        
        queue.add(startNode);
        
        while(!queue.isEmpty()) {
            Node curNode = queue.poll();

            for(Node node : graph[curNode.number]) {
                if(curNode.weight + node.weight < weights[node.number]) {
                    weights[node.number] = curNode.weight + node.weight;
                    queue.add(new Node(node.number, weights[node.number]));
                }
            }
        }
        
        return weights;
    }
}

File: 프로그래머스/2/12978. 배달/배달.py
import heapq

graph = []

class Node:
    def __init__(self, dest, weight):
        self.dest = dest
        self.weight = weight
        
    def __lt__(self, other):
        return self.weight < other.weight

def solution(N, road, K):
    answer = 0
    graph = [[] for i in range(N+1)]
    
    for row in road:
        src = row[0]
        dest = row[1]
        weight = row[2]
        graph[src].append((dest, weight))
        graph[dest].append((src,weight))
        
    queue = []
    distance = [-1 for i in range(N+1)]
    heapq.heappush(queue, Node(1, 0))
    distance[1] = 0
    
    
    while queue:
        cur = heapq.heappop(queue)
        # if distance[cur.dest] != -1 and cur.weight >= distance[cur.dest]:
        #     continue
            
        nexts = graph[cur.dest]
        for next in nexts:
            next_dist = cur.weight + next[1]
            if distance[next[0]] == -1 or next_dist < distance[next[0]]:
                distance[next[0]] = next_dist
                heapq.heappush(queue, Node(next[0], next_dist))
    
    print(distance)
    count = sum(1 for x in distance if x >= 0 and x <= K)
    
    return count